### PR TITLE
Add CheckReturnValue annotation

### PR DESCRIPTION
**Pull request by vanniktech**
First draft on this.

So far I have only annotated `subscribeWith()` and `test()` methods.

I checked this against Error Prone and their [check](http://errorprone.info/bugpattern/CheckReturnValue) is able of picking up the RxJava annotation.

How do you feel if I extend `BaseTypeAnnotations` to test for `@CheckReturnValue` too?

Related #4878

### CODE REVIEW

File: src/main/java/io/reactivex/annotations/CheckReturnValue.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Marks methods whose return values should be checked.
+ * 
+ * @since 2.0.2 - experimental
+ */
+@Retention(RetentionPolicy.RUNTIME)
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/annotations/CheckReturnValue.java:27]: nit: static import makes it a little bit better

File: src/main/java/io/reactivex/annotations/CheckReturnValue.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+@Documented
+@Target(ElementType.METHOD)
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/annotations/CheckReturnValue.java:29]: And here

**Review by artem-zinnatullin — APPROVED** [review]
Gooooooood. Afaik Findbugs also able to process annotation with this name correctly.

File: src/main/java/io/reactivex/annotations/CheckReturnValue.java (hunk 1 above)
Comment (vanniktech) [reply src/main/java/io/reactivex/annotations/CheckReturnValue.java:27]: Yes I know however for RxJava it's common to not static import them (look at BackpressureSupport & SchedulerSupport) so I left it that way. I leave the final decision up to David.

File: src/main/java/io/reactivex/annotations/CheckReturnValue.java (hunk 2 above)
Comment (vanniktech) [reply src/main/java/io/reactivex/annotations/CheckReturnValue.java:29]: same as above

**Comment by vanniktech** [comment]
I quickly checked findbugs and it does not seem like they pick up our annotation since they have their [own](http://findbugs.sourceforge.net/api/edu/umd/cs/findbugs/annotations/CheckReturnValue.html)

File: src/main/java/io/reactivex/annotations/CheckReturnValue.java (hunk 1 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/annotations/CheckReturnValue.java:27]: It's fine as is.

**Review by akarnokd — APPROVED** [review]

**Comment by akarnokd** [comment]
Would you manually ignore methods like `Disposable subscribe()`?

**Comment by artem-zinnatullin** [comment]
@vanniktech see https://github.com/findbugsproject/findbugs/blob/264ae7baf890d2b347d91805c90057062b5dcb1e/findbugs/src/java/edu/umd/cs/findbugs/detect/BuildCheckReturnAnnotationDatabase.java#L120

**Comment by vanniktech** [comment]
@artem-zinnatullin weird somehow findbugs didn't pick it up though. Could be a configuration error on my side.

I'd argue that not annotating `Diposable subscribe()` is fine however I feel like `Disposable subscribe(Consumer<T>)` and the others should be annotated. I personally almost never use subscribeWith and prefer the variations with lambdas.

We're on the same point that static factory methods and operators such as flatMap, first(), last() etc should be annotated right? If so I'd apply those and then we can see which might need an annotation too.

**Comment by vanniktech** [comment]
@akarnokd added the `@CheckReturnValue` annotation to all public API methods of `Flowable`, `Maybe`, `Single`, `Observable` and `Completable` except for private methods, void methods and the `Disposable subscribe()` method. Is that okay now?

**Review by akarnokd — APPROVED** [review]

**Comment by akarnokd** [comment]
Yes.

/cc @JakeWharton

**Pull request merged**